packages/opentelemetry-plugin-redis/src/redis.ts

```diff
@@ -0,0 +1,192 @@
+/*!
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { BasePlugin } from '@opentelemetry/core';
+import * as redisTypes from 'redis';
+import * as shimmer from 'shimmer';
+import { SpanKind, CanonicalCode } from '@opentelemetry/types';
+import { AttributeNames } from './enums';
+import {
+  RedisPluginStreamTypes,
+  RedisPluginClientTypes,
+  RedisCommand,
+} from './types';
+import { EventEmitter } from 'events';
+
+export class RedisPlugin extends BasePlugin<typeof redisTypes> {
+  static readonly COMPONENT = 'redis';
+  readonly supportedVersions = ['^2.6.0']; // should be >= 2.6.0
+
+  constructor(readonly moduleName: string) {
+    super();
+  }
+
+  protected patch() {
+    if (this._moduleExports.RedisClient) {
+      this._logger.debug(
+        'Patching redis.RedisClient.prototype.internal_send_command'
+      );
+      shimmer.wrap(
+        this._moduleExports.RedisClient.prototype,
+        'internal_send_command',
+        this._getPatchInternalSendCommand()
+      );
+
+      this._logger.debug('patching redis.create_stream');
+      shimmer.wrap(
+        this._moduleExports.RedisClient.prototype,
+        'create_stream',
+        this._getPatchCreateStream()
+      );
+
+      this._logger.debug('patching redis.createClient');
+      shimmer.wrap(
+        this._moduleExports,
+        'createClient',
+        this._getPatchCreateClient()
+      );
+    }
+    return this._moduleExports;
+  }
+
+  protected unpatch(): void {
+    if (this._moduleExports) {
+      shimmer.unwrap(
+        this._moduleExports.RedisClient.prototype,
+        'internal_send_command'
+      );
+      shimmer.unwrap(
+        this._moduleExports.RedisClient.prototype,
+        'create_stream'
+      );
+      shimmer.unwrap(this._moduleExports, 'createClient');
+    }
+  }
+
+  /**
+   * Patch internal_send_command(...) to trace requests
+   */
+  private _getPatchInternalSendCommand() {
+    const plugin = this;
+    return function internal_send_command(original: Function) {
+      return function internal_send_command_trace(
```

Comment:
Would it make sense to move this function out of the class and pass in `plugin` as an argument (would need to wrap with an arrow function)? That could reducer the indentation of it

Reply:
Is this what you had in mind? 8654a71